lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
+                        stopException, Level.Error, "Exception thrown on Task Stop", LOGGER);
                 }
+
+                Heartbeat();
             }
         }
 
         public void SetResult(byte[] result)
         {
             lock (_heartBeatManager)
             {
+                if (_state == TaskState.Failed)
+                {
+                    if (!_lastException.IsPresent())
+                    {
+                        throw new ApplicationException("Expected failed Task to already have set an Exception.");
```

Comment:
Looks like this is optional in SetException...do we really want to generate new exception here?